### PR TITLE
Specify column types returned by lcd

### LINKED ISSUE #396: lcd returning different vector types
Hi,
Thanks for the great package...
I'm running into a minor issue where `lcd()` sometimes returns a character column and sometimes returns a numeric column.  Reprex below:

```r
library(rnoaa)
x.1 <- lcd(station = "74746003904", year = 2010)
x.2 <- lcd(station = "74746003904", year = 2011)

str(x.1$source)
```

```
chr [1:11154] "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" "4" ...
```

```r
str(x.2$source)
```

```
int [1:11278] 7 7 7 7 7 7 7 7 7 7 ...
```

In the `echor` package which downloads various EPA environmental permit data, we ended up defining all the column classes as character when reading in the downloaded csv/json files to facilitate bulk data downloading (which is essentially what I'm using `rnoaa` for). I recognize that might not be a desired solution so completely understand if this issue is closed as is. 

## Session Info
```
R version 4.0.5 (2021-03-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19043)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tidyr_1.1.3        purrr_0.3.4        dplyr_1.0.5        readr_1.4.0        rnoaahelpers_0.1.0 rnoaa_1.3.4       

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6        pillar_1.6.0      compiler_4.0.5    prettyunits_1.1.1 progress_1.2.2    tools_4.0.5      
 [7] digest_0.6.27     jsonlite_1.7.2    lubridate_1.7.10  lifecycle_1.0.0   tibble_3.1.1      gtable_0.3.0     
[13] pkgconfig_2.0.3   rlang_0.4.11      DBI_1.1.1         cli_2.5.0         rstudioapi_0.13   crul_1.1.0       
[19] curl_4.3.1        gridExtra_2.3     withr_2.4.2       xml2_1.3.2        fauxpas_0.5.0     hms_1.0.0        
[25] rappdirs_0.3.3    generics_0.1.0    vctrs_0.3.8       triebeard_0.3.0   rprojroot_2.0.2   grid_4.0.5       
[31] prompt_1.0.1      tidyselect_1.1.1  data.table_1.14.0 glue_1.4.2        httpcode_0.3.0    here_1.0.1       
[37] R6_2.5.0          fansi_0.4.2       XML_3.99-0.6      sessioninfo_1.1.1 whisker_0.4       hoardr_0.5.2     
[43] ggplot2_3.3.3     magrittr_2.0.1    urltools_1.7.3    scales_1.1.1      ps_1.6.0          ellipsis_0.3.2   
[49] assertthat_0.2.1  colorspace_2.0-0  utf8_1.2.1        munsell_0.5.0     crayon_1.4.1    
```

### PR DESCRIPTION
**Pull request by mps9506**
## Description
`lcd()` returns consistent types for each column. New internal function, `lcd_columns()` returns named character vector (accessible and changable by users) for use by a new argument, `col_types`, in `lcd()`. This is ultimately used by `safe_read_csv()` and the `colClasses` argument in `data.table::fread()`. This is slightly more complicated than the initial code used to ensure types are returned consistently. However, this allows users to specify the type if required or if I specified something incorrectly as a default.  I think having the types stored in `lcd_columns()` will make updating in the future easier if new variables are added or changed in the future. `fread` will revert the column type to a lower type (with message) if an incompatible type is specified.

## Related Issue
closes #396 

## Example
```r
x.1 <- lcd(station = "74746003904", year = 2010)
x.1[c(1:6)]

# A tibble: 11,154 x 6
   station     date                latitude longitude elevation name                                   
   <chr>       <dttm>                 <dbl>     <dbl>     <dbl> <chr>                                  
 1 74746003904 2010-01-01 00:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 2 74746003904 2010-01-01 01:24:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 3 74746003904 2010-01-01 01:51:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 4 74746003904 2010-01-01 01:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 5 74746003904 2010-01-01 02:35:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 6 74746003904 2010-01-01 02:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 7 74746003904 2010-01-01 03:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 8 74746003904 2010-01-01 04:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 9 74746003904 2010-01-01 05:27:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
10 74746003904 2010-01-01 05:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
# ... with 11,144 more rows



col_types <- lcd_columns(STATION = "numeric")
x.2 <- lcd(station = "74746003904", year = 2010, col_types = col_types)
x.2[c(1:6)]

# A tibble: 11,154 x 6
       station date                latitude longitude elevation name                                   
         <dbl> <dttm>                 <dbl>     <dbl>     <dbl> <chr>                                  
 1 74746003904 2010-01-01 00:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 2 74746003904 2010-01-01 01:24:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 3 74746003904 2010-01-01 01:51:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 4 74746003904 2010-01-01 01:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 5 74746003904 2010-01-01 02:35:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 6 74746003904 2010-01-01 02:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 7 74746003904 2010-01-01 03:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 8 74746003904 2010-01-01 04:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 9 74746003904 2010-01-01 05:27:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
10 74746003904 2010-01-01 05:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
# ... with 11,144 more rows
```

### CODE REVIEW

File: R/lcd.R [hunk 1]
```diff
@@ -44,13 +48,18 @@
...
   assert(year, c("character", "numeric", "integer"))
   assert_range(year, 1901:format(Sys.Date(), "%Y"))
 
   path <- lcd_get(station = station, year = year, ...)
-  tmp <- safe_read_csv(path)
+  
+  if (is.null(col_types)) {
+    col_types <- lcd_columns()
+  }
+  assert(col_types, c("character"))
```
Comment (sckott) [inline R/lcd.R:61]: In addition to checking col_types is character, it seems like you can run into trouble if the input to `col_types` is not of the correct length. e.g, 

```r
# using letters vector, length 26
lcd(station = "74746003904", year = 2010, col_types = letters)
#> Error in safe_read_csv(path, col_types = col_types) :
#>   file ~/Library/Caches/R/noaa_lcd/2010_74746003904.csv malformed; delete file and try again
```

File: R/lcd.R [hunk 2]
```diff
@@ -95,3 +104,277 @@ lcd_base <- function() {
...
+  WindEquipmentChangeDate="character") {
+  # setting the column types for data as described in:
+  # https://www1.ncdc.noaa.gov/pub/data/ish/ish-format-document.pdf
+  # and: https://www.ncei.noaa.gov/data/local-climatological-data/doc/LCD_documentation.pdf
+  # daily and monthly values might include special indicators with the value or
+  # instead of a value
+  # s = suspect value, T = Trace, M = missing, * extreme max or min for month
+  # to retain this information, values are kept as character for user to handle
+  argg <- unlist(c(as.list(environment())))
+  return(argg)
```
Comment (sckott) [inline R/lcd.R:379]: seems like we should check here that the input values to each parameter is a proper R class, e.g., `lcd_columns(STATION = 5)` doesn't make much sense. although, fread does give a nice error message at least when you use it in `lcd`

```r
lcd(station = "74746003904", year = 2010, col_types = lcd_columns(STATION = 5))
#> Error in safe_read_csv(path, col_types = col_types) :
#>   Column 'STATION' was requested to be '5' but fread encountered the following error:
#> 	no method or default for coercing “character” to “5”
#> so the column has been left as type 'character'
```

**Review by sckott — CHANGES_REQUESTED** [review]
Nice work! A few comments inline

File: R/lcd.R (hunk 2 above)
Comment (mps9506) [reply R/lcd.R:379]: Agreed. I've added a function to zzz.r that returns an error message if wrong types are entered. This is run by both `lcd()` and `lcd_columns()` with some logic to reduce the chance it is run twice.

```r
## works
x.1 <- lcd(station = "74746003904", year = 2010, col_types = c(STATION="character"))
x.1[c(1:6)]

# A tibble: 11,154 x 6
   station     date                latitude longitude elevation name                                   
   <chr>       <dttm>                 <dbl>     <dbl>     <dbl> <chr>                                  
 1 74746003904 2010-01-01 00:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 2 74746003904 2010-01-01 01:24:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 3 74746003904 2010-01-01 01:51:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 4 74746003904 2010-01-01 01:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 5 74746003904 2010-01-01 02:35:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 6 74746003904 2010-01-01 02:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 7 74746003904 2010-01-01 03:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 8 74746003904 2010-01-01 04:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 9 74746003904 2010-01-01 05:27:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
10 74746003904 2010-01-01 05:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
# ... with 11,144 more rows
```

```r 
## works
x.2 <- lcd(station = "74746003904", year = 2010, col_types = lcd_columns(DATE = "character"))
x.2[c(1:6)]
# A tibble: 11,154 x 6
   station     date                latitude longitude elevation name                                   
   <chr>       <chr>                  <dbl>     <dbl>     <dbl> <chr>                                  
 1 74746003904 2010-01-01T00:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 2 74746003904 2010-01-01T01:24:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 3 74746003904 2010-01-01T01:51:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 4 74746003904 2010-01-01T01:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 5 74746003904 2010-01-01T02:35:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 6 74746003904 2010-01-01T02:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 7 74746003904 2010-01-01T03:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 8 74746003904 2010-01-01T04:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
 9 74746003904 2010-01-01T05:27:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
10 74746003904 2010-01-01T05:53:00     30.6     -96.4        93 COLLEGE STATION EASTERWOOD FIELD, TX US
# ... with 11,144 more rows
```

```r
## expect errors
col_types <- c(STATION=5)
lcd(station = "74746003904", year = 2010, col_types = col_types)

Error in lcd(station = "74746003904", year = 2010, col_types = col_types) : 
  STATION must equal a valid R class ('character', 'integer', 'numeric', 'factor', 'integer64', 'POSIXct')
```

``r
## expect errors
lcd_columns(STATION = 5)

Error in lcd_columns(STATION = 5) : 
  STATION must equal a valid R class ('character', 'integer', 'numeric', 'factor', 'integer64', 'POSIXct')
```

Hopefully these are useful error messages for users.

File: R/lcd.R (hunk 1 above)
Comment (mps9506) [reply R/lcd.R:61]: I think for the purpose of this function, we can specify that a named vector is expected, then the length of the vector does not matter. If the column name is misspecified by the user, `fread()`  returns a useful error that the column name was not found.
```r
# expect error
x <- letters
lcd(station = "74746003904", year = 2010, col_types = x)

Error in lcd(station = "74746003904", year = 2010, col_types = x) : 
  col_types must be a named vector, see lcd_columns() for an example and expected names
```

```r
# expect a useful error
names(x) <- x
lcd(station = "74746003904", year = 2010, col_types = x)

Error in lcd(station = "74746003904", year = 2010, col_types = x) : 
  a must equal a valid R class ('character', 'integer', 'numeric', 'factor', 'integer64', 'POSIXct')
b must equal a valid R class ('character', 'integer', 'numeric', 'factor', 'integer64', 'POSIXct')
c must equal a valid R class ('character', 'integer', 'numeric', 'factor', 'integer64', 'POSIXct')
d must equal a valid R class ('character', 'integer', 'numeric', 'factor', 'integer64', 'POSIXct')
```

```r
# error for wrong column name
lcd(station = "74746003904", year = 2010, col_types = c(station = "character"))

 Error in safe_read_csv(path, col_types = col_types) : 
  Column name 'station' (colClasses[[1]][1]) not found 
```

Comment (sckott) [reply R/lcd.R:72]: sounds good, thanks

File: R/lcd.R (hunk 2 above)
Comment (sckott) [reply R/lcd.R:395]: Looks great

**Review by sckott — APPROVED** [review]

**Pull request merged**